fog/test-client/src/test_client.rs

```diff
@@ -781,3 +864,125 @@ impl Drop for ReceiveTxWorker {
+}
+
+impl TxInfo {
+    /// Clear the TxInfo
+    pub fn clear(&self) {
+        *self.tx.lock().unwrap() = None;
+        *self.tx_propose_block_count.lock().unwrap() = None;
+        *self.tx_appeared.lock().unwrap() = None;
+    }
+
+    /// Set the Tx that we are sending (immediately after it is built)
+    pub fn set_tx(&self, tx: &Tx) {
+        *self.tx.lock().unwrap() = Some(tx.clone());
+    }
+
+    /// Set the block count returned by tx_propose (immediately after it is
+    /// known)
+    pub fn set_tx_propose_block_count(&self, count: u64) {
+        *self.tx_propose_block_count.lock().unwrap() = Some(count);
+    }
+
+    /// Set the index in which the tx appeared (immediately after it is known)
+    pub fn set_tx_appeared_block_index(&self, index: BlockIndex) {
+        *self.tx_appeared.lock().unwrap() = Some(index);
+    }
+}
+
+impl core::fmt::Display for TxInfo {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        self.tx_propose_block_count
+            .lock()
+            .unwrap()
+            .map(|proposed| {
+                write!(
+                    f,
+                    "Proposed at block index ~{}, ",
+                    proposed.saturating_sub(1)
+                )
+            })
+            .transpose()?;
+        self.tx_appeared
+            .lock()
+            .unwrap()
+            .map(|appeared| write!(f, "Appeared in block index {}, ", appeared))
+            .transpose()?;
+        self.tx
+            .lock()
+            .unwrap()
+            .as_ref()
+            .map(|tx| {
+                write!(
+                    f,
+                    "TxOut public keys: [{}]",
+                    HexList(tx.prefix.outputs.iter().map(|x| x.public_key.as_bytes()))
+                )
+            })
+            .transpose()?;
+        Ok(())
+    }
+}
+
+/// Object which helps tracking the healthy status
+#[derive(Default)]
+pub struct HealthyTracker {
+    // Set to i for the duration of the i'th transfer
+    counter: AtomicUsize,
+    // Whether we have observed any failure
+    have_failure: AtomicBool,
+    // The counter value during the most recent failure
+    last_failure: AtomicUsize,
+    // The number of clients
+    num_clients: usize,
+}
+
+impl HealthyTracker {
+    /// Make a new healthy tracker.
+    /// Sets IS_HEALTHY to true initially.
+    /// Takes num_clients which is the number of successful transfers before we
+    /// consider ourselves healthy again
+    pub fn new(num_clients: usize) -> Self {
+        counters::IS_HEALTHY.set(1);
+        Self {
+            num_clients,
+            ..Default::default()
+        }
+    }
+
+    /// Set the counter value, and maybe update healthy status
+    pub fn set_counter(&self, counter: usize) {
+        self.counter.store(counter, Ordering::SeqCst);
+        // If:
+        // * there is a failure
+        // * the failure happened at least num_clients ago
+        // then set ourselves healthy again
```

Comment:
Consider add an explanation on why this logic makes sense and not just explaining what it does.